### PR TITLE
tests: Add Windows threads to stress_mt

### PR DESCRIPTION
**Pull request by tormodvolden**
Build the stress_mt test on all platforms.

Different from PR #1246, this uses Windows threads and not Posix threads on Windows.

### CODE REVIEW

**Comment by mcuee** [comment]
Nice one. Tested with VS2022 on Windows 11 64bit Intel.

```
PS C:\work\libusb\libusb_pr1252> msbuild -m -v:m -p:PlatformToolset=v143,Platform=x64,Configuration=Release .\msvc\libusb.sln
Microsoft (R) Build Engine version 17.1.0+ae57d105c for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

  core.c
  descriptor.c
  events_windows.c
  hotplug.c
  io.c
  strerror.c
  sync.c
  getopt.c
  threads_windows.c
  windows_common.c
  getopt1.c
  core.c
  windows_usbdk.c
  getopt.vcxproj -> C:\work\libusb\libusb_pr1252\build\v143\x64\Release\getopt.lib
  descriptor.c
  windows_winusb.c
  libusb_static.vcxproj -> C:\work\libusb\libusb_pr1252\build\v143\x64\Release\lib\libusb-1.0.lib
  events_windows.c
  stress.c
  hotplug.c
  sam3u_benchmark.c
  io.c
  testlib.c
  dpfp.c
  hotplugtest.c
  Generating code
  Generating code
  Generating code
  Generating code
  strerror.c
  sync.c
  Finished generating code
  threads_windows.c
  testlibusb.c
  xusb.c
  dpfp.c
  listdevs.c
  sam3u_benchmark.vcxproj -> C:\work\libusb\libusb_pr1252\build\v143\x64\Release\sam3u_benchmark.exe
  Generating code
  Generating code
  Generating code
  Generating code
  windows_common.c
  Finished generating code
  windows_usbdk.c
  stress_mt.c
  testlibusb.vcxproj -> C:\work\libusb\libusb_pr1252\build\v143\x64\Release\testlibusb.exe
  Finished generating code
  Finished generating code
  Generating code
  Finished generating code
  ezusb.c
  Finished generating code
  windows_winusb.c
  dpfp_threaded.vcxproj -> C:\work\libusb\libusb_pr1252\build\v143\x64\Release\dpfp_threaded.exe
  stress_mt.vcxproj -> C:\work\libusb\libusb_pr1252\build\v143\x64\Release\stress_mt.exe
  hotplugtest.vcxproj -> C:\work\libusb\libusb_pr1252\build\v143\x64\Release\hotplugtest.exe
  fxload.c
  Finished generating code
  Generating code
     Creating library C:\work\libusb\libusb_pr1252\build\v143\x64\Release\libusb_dll\..\dll\libusb-1.0.lib and object C
  :\work\libusb\libusb_pr1252\build\v143\x64\Release\libusb_dll\..\dll\libusb-1.0.exp
  Generating code
  Finished generating code
  Finished generating code
  Finished generating code
  stress.vcxproj -> C:\work\libusb\libusb_pr1252\build\v143\x64\Release\stress.exe
  dpfp.vcxproj -> C:\work\libusb\libusb_pr1252\build\v143\x64\Release\dpfp.exe
  xusb.vcxproj -> C:\work\libusb\libusb_pr1252\build\v143\x64\Release\xusb.exe
  listdevs.vcxproj -> C:\work\libusb\libusb_pr1252\build\v143\x64\Release\listdevs.exe
  fxload.vcxproj -> C:\work\libusb\libusb_pr1252\build\v143\x64\Release\fxload.exe
  Finished generating code
  libusb_dll.vcxproj -> C:\work\libusb\libusb_pr1252\build\v143\x64\Release\dll\libusb-1.0.dll

PS C:\work\libusb\libusb_pr1252> .\build\v143\x64\Release\stress_mt.exe
Starting multithreaded init and exit test...
Thread 0 started
Thread 1 started
Thread 3 started
Thread 4 started
Thread 5 started
Thread 2 started
Thread 6 started
Thread 7 started
Thread 0 done
Thread 1 done
Thread 6 done
Thread 3 done
Thread 4 done
Thread 2 done
Thread 5 done
Thread 7 done
All Done


```

**Comment by mcuee** [comment]
No problems with MSYS2 mingw64 as well.

```
MINGW64 /c/work/libusb/libusb_pr1252
$ ./tests/stress_mt.exe
Starting multithreaded init and exit test...
Thread 0 started
Thread 3 started
Thread 4 started
Thread 2 started
Thread 5 started
Thread 1 started
Thread 6 started
Thread 7 started
Thread 6 done
Thread 1 done
Thread 5 done
Thread 3 done
Thread 7 done
Thread 2 done
Thread 4 done
Thread 0 done
All Done

```

File: tests/stress_mt.c [hunk 1]
```diff
@@ -1,15 +1,85 @@
...
 #include <libusb.h>
 #include <stdio.h>
+
+#if defined(PLATFORM_POSIX)
+
 #include <pthread.h>
+typedef pthread_t thread_t;
+typedef void * thread_return_t;
+#define THREAD_RETURN_VALUE NULL
+#define STDCALL
```
Comment (Youw) [inline tests/stress_mt.c:31]: I'd name this macro `THREAD_CALL_TYPE` (or similar) as at the place of the usage one might mistakenly think that someone tries to have an `STDCALL` function is doesn't make sense in cross-platform code.

**Review by Youw — APPROVED** [review]
The change actually looks a lot simpler than I thought it would.

**Comment by tormodvolden** [comment]
The third commit is not about Windows threads of course, but just adding more test coverage.

Note the MSVC project changes were done manually (with vi and uuidgen) but it seems it works fine in the Appveyor CI. Maybe someone with MSVC can open the project there and check that everything is fine.

**Comment by tormodvolden** [comment]
The third commit is broken on 64-bit MSVC.

**Comment by mcuee** [comment]
Looks good. Tested with Visual Studio 2022 under Windows 11.

```
PS C:\work\libusb\libusb_pr1252_v2\msvc> ..\build\v143\x64\Release\stress_mt.exe
Running multithreaded init/exit test...
Starting 8 threads
Running multithreaded init/exit test with enumeration...
Starting 8 threads
Thread 0 discovered 9 devices
Thread 1 discovered 9 devices
Thread 2 discovered 9 devices
Thread 3 discovered 9 devices
Thread 4 discovered 9 devices
Thread 5 discovered 9 devices
Thread 6 discovered 9 devices
Thread 7 discovered 9 devices
All done, 0 errors

```

**Comment by mcuee** [comment]
> The third commit is not about Windows threads of course, but just adding more test coverage.
> 
> Note the MSVC project changes were done manually (with vi and uuidgen) but it seems it works fine in the Appveyor CI. Maybe someone with MSVC can open the project there and check that everything is fine.

No issues with opening the project file using Visual Studio 2022. 

Build log for reference.

```
Build started...
1>------ Build started: Project: libusb_static, Configuration: Release-MT x64 ------
1>core.c
1>descriptor.c
1>events_windows.c
1>hotplug.c
1>io.c
1>strerror.c
1>sync.c
1>threads_windows.c
1>windows_common.c
1>windows_usbdk.c
1>windows_winusb.c
1>libusb_static.vcxproj -> C:\work\libusb\libusb_pr1252\build\v142\x64\Release-MT\lib\libusb-1.0.lib
2>------ Build started: Project: stress_mt, Configuration: Release x64 ------
2>stress_mt.c
2>Generating code
2>Finished generating code
2>stress_mt.vcxproj -> C:\work\libusb\libusb_pr1252\build\v142\x64\Release\stress_mt.exe
========== Build: 2 succeeded, 0 failed, 0 up-to-date, 0 skipped ==========

```

**Comment by mcuee** [comment]
> Hmm, libusb_get_device_list() return ssize_t, and on error LIBUSB_ERROR_NO_MEM which is -11. Actually the ssize_t type is only guaranteed to include -1 on the negative side. I wonder if the type was chosen in case some people have more than MAX_INT devices connected?

Yes that is a interesting. I am not so sure why libusb_get_device_list() return ssize_t.

**Comment by mcuee** [comment]
Just a test with PR #1202 integrated into this PR. It seems to work as well.

```
PS C:\work\libusb\libusb_pr1252_pr1202\msvc> ..\build\v142\x64\Release\stress_mt.exe
Running multithreaded init/exit test...
Starting 8 threads
Running multithreaded init/exit test with enumeration...
Starting 8 threads
Thread 0 discovered 22 devices
Thread 1 discovered 22 devices
Thread 2 discovered 22 devices
Thread 3 discovered 22 devices
Thread 4 discovered 22 devices
Thread 5 discovered 22 devices
Thread 6 discovered 22 devices
Thread 7 discovered 22 devices
All done, 0 errors

```

**Comment by mcuee** [comment]
@tormodvolden 

I found something strange that the context is always 0 under Windows.

```
PS C:\work\libusb\libusb_pr1252> git diff
diff --git a/tests/stress_mt.c b/tests/stress_mt.c
index 07a9a8b2..8785c945 100644
--- a/tests/stress_mt.c
+++ b/tests/stress_mt.c
@@ -108,6 +108,7 @@ static thread_return_t THREAD_CALL_TYPE init_and_exit(void * arg)
                        ti->devcount = cnt;
                        libusb_free_device_list(devs, 1);
                }
+               printf("Thread %d found %lld of devices in context %d \n", ti->number, ti->devcount, r);

                libusb_exit(ctx);
        }

PS C:\work\libusb\libusb_pr1252> .\build\v142\x64\Release\stress_mt.exe >stress_mt.log 2>&1
```

[stress_mt.zip](https://github.com/libusb/libusb/files/10846213/stress_mt.zip)

**Comment by tormodvolden** [comment]
r is the return value, not the context.

**Comment by mcuee** [comment]
> r is the return value, not the context.

Oops, I forgot that.

**Comment by mcuee** [comment]
@tormodvolden 

I think this PR is good to go now.

**Comment by tormodvolden** [comment]
The first commits can safely be merged. The third is still getting updates :) but should be fine for now. It would also be nice to verify that the original issues we once had in the library would still be found by the updated test.

**Comment by mcuee** [comment]
More tests under Windows.

```
PS C:\work\libusb\libusb_pr1252> cat .\run_10times.ps1
for ($i=1; $i -le 10; $i++) {echo $i; .\build\v142\x64\Release\stress_mt.exe}

PS C:\work\libusb\libusb_pr1252> Measure-Command {.\run_10times.ps1}
Days              : 0
Hours             : 0
Minutes           : 1
Seconds           : 48
Milliseconds      : 123
Ticks             : 1081233190
TotalDays         : 0.00125142730324074
TotalHours        : 0.0300342552777778
TotalMinutes      : 1.80205531666667
TotalSeconds      : 108.123319
TotalMilliseconds : 108123.319

```

**Comment by tormodvolden** [comment]
Good, it is also a benchmark of the enumeration. I see each test run takes ~11 seconds on your machine with 23 devices.

**Comment by mcuee** [comment]
> Good, it is also a benchmark of the enumeration. I see each test run takes ~11 seconds on your machine with 23 devices.

Yes, this machine is a bit more complicated with docking station and external hub.

```
MINGW64 /c/work/libusb/libusb_pr1252
$ time ./examples/testlibusb.exe
Dev (bus 3, device 5): 0424 - 5744 speed: 5G
Dev (bus 3, device 6): 0BDA - 5413 speed: 480M
Dev (bus 3, device 2): 0BDA - 0413 speed: 5G
Dev (bus 3, device 17): 0424 - 2740 speed: 480M
  Manufacturer:              Microchip Tech
  Product:                   Hub Controller
Dev (bus 3, device 7): 413C - B06E speed: 480M
Dev (bus 1, device 6): 8087 - 0AAA speed: 12M
Dev (bus 3, device 8): 0424 - 2744 speed: 480M
Dev (bus 3, device 4): 0BDA - 5487 speed: 480M
Dev (bus 1, device 9): 0403 - 6001 speed: 12M
Dev (bus 3, device 1): 0BDA - 0487 speed: 10G
Dev (bus 1, device 2): 2717 - 5013 speed: 12M
  Manufacturer:              MI
  Product:                   Mi Wireless Mouse
Dev (bus 3, device 15): 0BDA - 402E speed: 480M
Dev (bus 1, device 0): 8086 - A36D speed: 5G
Dev (bus 3, device 12): 413C - 2107 speed: 1.5M
  Manufacturer:              DELL
  Product:                   Dell USB Entry Keyboard
Dev (bus 1, device 4): 0A5C - 5842 speed: 480M
Dev (bus 3, device 0): 8086 - 15F0 speed: 10G
Dev (bus 3, device 16): 1679 - 2001 speed: 480M
  Manufacturer:              Total Phase, Inc.
  Product:                   Beagle I2C/SPI/MDIO Protocol Analyzer
Dev (bus 2, device 0): 8086 - 15DB speed: 5G
Dev (bus 3, device 3): 0BDA - 8153 speed: 5G
Dev (bus 3, device 9): 413C - B06F speed: 480M
Dev (bus 3, device 13): 046D - C077 speed: 1.5M
  Manufacturer:              Logitech
  Product:                   USB Optical Mouse
Dev (bus 1, device 5): 0BDA - 58FD speed: 480M

real    0m0.827s
user    0m0.000s
sys     0m0.031s

$ time ./examples/listdevs.exe
0424:5744 (bus 3, device 5) path: 4.3.2
0bda:5413 (bus 3, device 6) path: 2.3
0bda:0413 (bus 3, device 2) path: 4.3
0424:2740 (bus 3, device 17) path: 2.3.2.5
413c:b06e (bus 3, device 7) path: 2.5
8087:0aaa (bus 1, device 6) path: 14
0424:2744 (bus 3, device 8) path: 2.3.2
0bda:5487 (bus 3, device 4) path: 2
0403:6001 (bus 1, device 9) path: 2
0bda:0487 (bus 3, device 1) path: 4
2717:5013 (bus 1, device 2) path: 1
0bda:402e (bus 3, device 15) path: 2.3.4
8086:a36d (bus 1, device 0)
413c:2107 (bus 3, device 12) path: 2.3.2.3
0a5c:5842 (bus 1, device 4) path: 10
8086:15f0 (bus 3, device 0)
1679:2001 (bus 3, device 16) path: 2.3.2.2
8086:15db (bus 2, device 0)
0bda:8153 (bus 3, device 3) path: 4.4
413c:b06f (bus 3, device 9) path: 2.3.5
046d:c077 (bus 3, device 13) path: 2.3.2.4
0bda:58fd (bus 1, device 5) path: 11

real    0m0.702s
user    0m0.000s
sys     0m0.046s

$ time ./tests/stress_mt.exe
Running multithreaded init/exit test...
Starting 8 threads
Running multithreaded init/exit test with enumeration...
Starting 8 threads
Thread 0 discovered 22 devices
Thread 1 discovered 22 devices
Thread 2 discovered 22 devices
Thread 3 discovered 22 devices
Thread 4 discovered 22 devices
Thread 5 discovered 22 devices
Thread 6 discovered 22 devices
Thread 7 discovered 22 devices
All done, 0 errors

real    0m11.327s
user    0m0.000s
sys     0m0.031s
```

**Comment by tormodvolden** [comment]
> It would also be nice to verify that the original issues we once had in the library would still be found by the updated test.

I tried simply reverting commit 4e2de881 (which fixed the original issue) on top of this but there were merge conflicts, and I cannot test on darwin right now anyway. The easiest way to test would be to checkout 4e2de881^ and cherry-pick the three commits here on top.

**Pull request closed without merging**